[SPARK-43529][SQL] Support general constant expressions as CREATE/REPLACE TABLE OPTIONS values

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR updates the SQL compiler to support general constnat expressions in the syntax for CREATE/REPLACE TABLE OPTIONS values, rather than restricting to a few types of literals only.

* The analyzer now checks that the provided expressions are in fact `foldable`, and throws an error message otherwise.
* This error message that users encounter in these cases improves from a general "syntax error at or near <location>" to instead indicate that the syntax is valid, but only constant expressions are supported in these contexts.

### Why are the changes needed?

This makes it easier to provide OPTIONS lists in SQL, supporting use cases like concatenating strings with `||`.

### Does this PR introduce _any_ user-facing change?

Yes, the SQL syntax changes.

### How was this patch tested?

This PR adds new unit test coverage.

dtenedor committed:
commit

commit

dtenedor committed:
fix build

dtenedor committed:
fix tests

dtenedor committed:
fix tests

dtenedor committed:
fix tests

dtenedor committed:
fix tests

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -3187,7 +3187,24 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
+        null
+      } else {
+        expression(property.value) match {
+          case Literal(str: UTF8String, StringType) => str.toString
+          case Literal(_, BooleanType) => property.value.getText.toLowerCase(Locale.ROOT)
+          case Literal(_, IntegerType | DecimalType()) => property.value.getText
```
Are there any reasons to support only literals but not foldable expressions?

dtenedor replied:
Yea we could support foldable expressions in the future. And constant-fold them, and then store the text of the resulting constant in the option value instead.

My idea was to separate concerns by doing just the parser change here first, but not doing this constant folding or changing the behavior yet. This PR only supports the same types of literal values for the option values as before the PR, simply improving error messages in other cases from generic `syntax error` to more descriptive messages indicating that only certain literals are supported. Then in another PR, we could implement the constant-folding.

But if you prefer, I can implement the constant-folding in this PR as well, please let me know.

gengliangwang replied:
Seems no new syntax is supported after this PR, while the title says support general expressions... I prefer having the constant folding in this PR

dtenedor replied:
OK, this is done.

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -3187,7 +3189,37 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
+          case Project(Seq(a: Alias), OneRowRelation()) => a.child
+        }.get
+        result match {
+          case expr if expr.foldable => expr.eval().toString
```
How about to wrap `expr` by the `Cast` to string expression and invoke `eval`. `toString` and casting to string can  output different results if we care of them.

dtenedor replied:
This is done.

dtenedor replied:
Update: I found that using `Cast` does not work in all cases, such as timestamps since we do not currently support casting timestamps to strings. I switched this back to use `.toString` and added a comment explaining this. It seems to work well for different types of expressions.

dtenedor committed:
respond to code review comments

dtenedor committed:
add another test

dtenedor committed:
add more test cases

dtenedor committed:
respond to code review comments

dtenedor commented:
Note @MaxGekk @gengliangwang it says the GitHub actions CI failed, but actually this was just an unrelated PySpark flake and they actually passed :)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -3187,7 +3189,42 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
+            ctx)
+        }
+        val plan = try {
+          val analyzer: Analyzer = ResolveDefaultColumns.DefaultColumnAnalyzer
```
@dtenedor It's too hacky to use the analyzer within the parser.  Could you verify it in the analyzer?

dtenedor replied:
Good point. I updated this to avoid using the analyzer, and just check if the expression is foldable and `eval` it instead. This supports all types of literals and some basic operations now. Function calls are not supported yet; I can support them later, but that change is more invasive.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @gengliangwang for your review! Please take another look.

dtenedor reviewed (commented):
Thanks @gengliangwang for your review! Please take another look.

dtenedor commented:
Update on this: I thought of a simpler way to break apart this work into smaller pieces so we can make gradual improvements. Let me make a commit and then I can ping this thread again.

dtenedor committed:
move constant folding to analysis

move constant folding to analysis

dtenedor committed:
defer constant folding to analysis

dtenedor commented:
This is done, I have deferred the constant folding logic to the analyzer. This is ready for another look.

gerashegalov commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,76 @@
+        case _: AnalysisException =>
+          throw optionNotConstantError(key)
+      }
+      val result: Expression = plan.collectFirst {
+        case Project(Seq(a: Alias), OneRowRelation()) => a.child
+      }.get
+      val newValue = result match {
+        case expr if expr.isInstanceOf[Literal] =>
+          // Note: we use 'toString' here instead of using a Cast expression to support some types
+          // of literals where casting to string is not supported.
+          expr.toString
+        case _ =>
+          throw optionNotConstantError(key)
+      }
```
L56 potentially risks an NoSuchElementException for None, or if not just for cleaner style I'd suggest

```suggestion
      val result: Option[Expression] = plan.collectFirst {
        case Project(Seq(a: Alias), OneRowRelation()) => a.child
      }
      val newValue = result match {
        case Some(expr@Literal(_, _)) =>
          // Note: we use 'toString' here instead of using a Cast expression to support some types
          // of literals where casting to string is not supported.
          expr.toString
        case _ =>
          throw optionNotConstantError(key)
      }
```

dtenedor replied:
Good idea, this is done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,76 @@
+ * commands such as CREATE TABLE. These expressions may be constant but non-literal, in which case
+ * we perform constant folding here.
+ */
+case class ResolveTableSpec(catalogManager: CatalogManager) extends Rule[LogicalPlan] {
```
I wonder if we can avoid creating a new rule for this.
Probably all we need is to make `mapExpressions` supports the type `Map[String, Expression]` https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala#L217
With that, analyzer will analyze the expression and the optimizer will do the constant folding.

dtenedor replied:
Sounds good, done!

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -1389,4 +1389,8 @@ case class TableSpec(
+    // If non-empty, this contains the values of the 'options' list as parsed but unresolved
+    // expressions. The analyzer will then constant-fold them to literals and clear this field,
+    // assigning the result to 'options'.
+    unresolvedOptions: Map[String, Option[Expression]] = Map.empty)
```
Why `Option[Expression]`? Can we make it `Map[String, Expression]`

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @gengliangwang for your review, I have responded to your suggestions.

dtenedor committed:
fix build

dtenedor committed:
fix build

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollection.scala:
```diff
@@ -92,6 +92,7 @@ object RuleIdCollection {
       "org.apache.spark.sql.catalyst.analysis.ResolveLateralColumnAliasReference" ::
       "org.apache.spark.sql.catalyst.analysis.ResolveOrderByAll" ::
       "org.apache.spark.sql.catalyst.analysis.ResolveRowLevelCommandAssignments" ::
+      "org.apache.spark.sql.catalyst.analysis.ResolveTableSpec" ::
```
Let remove this

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -1385,8 +1389,37 @@ case class DropIndex(
+    } else value match {
+      case Literal(litVal: String, _) => litVal
+      case Literal(litVal: UTF8String, _) => litVal.toString
+      case _ => value.sql
```
How can a non-literal expression convert back to itself with `toExpressionOptions`?

dtenedor replied:
I realize that in this case, the constant-folding step was not able to convert the option value to a literal, so it was non-constant; we return an appropriate error message in this case now.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -1385,8 +1389,49 @@ case class DropIndex(
-    external: Boolean)
+    external: Boolean) extends TreePatternBits {
+  /** This is a convenience method to obtain the options list as a map of strings to strings. */
+  lazy val options: Map[String, String] = {
```
This looks hacky. Ideally, a SQL query should work even when the optimizer rule `ConstantFolding`.

Sorry for getting back and forth: shall we have a `UnresolvedTableSpec` which contains `optionsList: Map[String, Expression]`. And another rule to resolve `UnresolvedTableSpec` as `TableSpec` if the expression in the map is all resolved? We can try using expression.eval() in the new rule.

dtenedor replied:
Sure, this is done.

dtenedor committed:
resolve merge conflicts

dtenedor committed:
fix appveyor build

dtenedor committed:
fix test

dtenedor committed:
resolve merge conflicts again

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,73 @@
+case class ResolveTableSpec(resolveOption: Expression => Expression) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.resolveOperatorsWithPruning(_.containsAnyPattern(COMMAND), ruleId) {
+      case t @ CreateTable(_, _, _, u: UnresolvedTableSpec, _) =>
```
nit: we can write this as 
```
case t: CreateTable if t.tableSpec.isInstanceOf[UnresolvedTableSpec]
```
so that we don't need to take care of the pattern match here when `CreateTable` is refactored. 
Same for the belows.

dtenedor replied:
Sounds good, done.

dtenedor committed:
respond to code review comments

dtenedor committed:
fix test

dtenedor committed:
fix test

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2406,6 +2407,15 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
+    val analyzed = executeSameContext(Project(Seq(Alias(expression, "col")()), OneRowRelation()))
+    val folded = ConstantFolding(analyzed)
+    folded match {
+      case Project(Seq(Alias(expression, _)), _) => expression
```
Do we need error handling if the pattern doesn't match?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,74 @@
+      case (key: String, value: Expression) =>
+        val newValue: String = try {
+          resolveOption(value) match {
+            case lit: Literal =>
```
Do we need error handling if the resolve result is not literal?

dtenedor replied:
Sure, the `catch` block  checked for `MatchError`, but I added a `case _` to make this explicit.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -1382,11 +1383,64 @@ case class DropIndex(
     serde: Option[SerdeInfo],
-    external: Boolean)
+    external: Boolean) extends TableSpec {
+  override def copy(
```
case class has built-in support for method copy. Why do we need this?

dtenedor replied:
I only added to copy the `TableSpec` with a new `location`. I switched this to just `withNewLocation` instead to simplify.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:
```diff
@@ -927,9 +927,14 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
+      val newOptionsList = Utils.redact(t.optionsExpressions.map { case (key, value) =>
+        (key, value.sql)
+      }).toMap
+      ResolvedTableSpec(
```
Why it becomes a `ResolvedTableSpec` here?

dtenedor replied:
I updated this to keep the object as an `UnresolvedTableSpec` here instead.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2406,6 +2407,15 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
   }
 
+  /** Helper method to resolve function calls within expressions of TableSpec options. */
+  def ResolveTableSpecOption(expression: Expression): Expression = {
```
If we manage to transform UnresolvedTableSpec in the method `mapExpressions`, probably we don't need this method.

dtenedor replied:
👍 I was able to remove this entirely, and `ResolveTableSpecOption` is no longer a separate analyzer rule. Instead, `ResolveSessionCatalog` now simply just resolves the `TableSpec` when needed directly.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Hi @gengliangwang I responded to your comments, please take another look. I think you will like this version, the analyzer now resolves the `TableSpec` directly when processing the CREATE/REPLACE TABLE command (rather than using a separate new analyzer rule).

dtenedor reviewed (commented):
Hi @gengliangwang I responded to your comments, please take another look. I think you will like this version, the analyzer now resolves the `TableSpec` directly when processing the CREATE/REPLACE TABLE command (rather than using a separate new analyzer rule).

dtenedor committed:
fix scalastyle

dtenedor committed:
fix scalastyle

dtenedor committed:
fix scalastyle

dtenedor committed:
fix scalastyle

dtenedor committed:
fix scalastyle

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -1382,11 +1407,59 @@ case class DropIndex(
+case class UnresolvedTableSpec(
+    properties: Map[String, String],
+    provider: Option[String],
+    optionsExpressions: Map[String, Expression],
```
@dtenedor I got a new idea about this.
How about we have a new expression for `optionsExpressions`
```
case class UnresolvedTableOptions(options: Seq[(String, Expression)])
  extends Expression with Unevaluable {
  override def nullable: Boolean = true

  override def dataType: DataType = MapType[StringType, StringType]

  override def children: Seq[Expression] = optionsExpressions.map(_._2)

  override protected def withNewChildrenInternal(
    newChildren: IndexedSeq[Expression]): Expression = {
      assert(options.length == newChildren.length)
      ...
      }
  }
}
```
So that we don't need to handle the treePatternBits or hacking the method `mapExpressions`. 
Note that here the Options is `Seq` instead of `Map` to make sure the expression order is fixed.

dtenedor replied:
Thanks for the good suggestion, this makes it simpler! This is done.

dtenedor committed:
respond to code review comments

dtenedor committed:
sync from master

dtenedor commented:
Hi @gengliangwang I implemented your suggestion, the code is simpler now, this is ready for another look.

dtenedor commented:
Note: the CI is actually passing, the pyspark failure is spurious/unrelated.

<img width="312" alt="image" src="https://github.com/apache/spark/assets/99207096/02176772-3e7c-4cbd-8308-297c6fd85066">

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -158,30 +158,37 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
     // For CREATE TABLE [AS SELECT], we should use the v1 command if the catalog is resolved to the
     // session catalog and the table provider is not v2.
-    case c @ CreateTable(ResolvedV1Identifier(ident), _, _, _, _) =>
+    case c @ CreateTable(ResolvedV1Identifier(ident), _, _, u: UnresolvedTableSpec, _,
```
QQ: why separate UnresolvedOptionsList from UnresolvedTableSpec? Now the rule has to match both UnresolvedTableSpec and UnresolvedOptionsList.

gengliangwang replied:
Nit: shall we have a single rule for resolving the table spec? This rule is for resolving session catalogs.

dtenedor replied:
Sure, I switched it back to using `ResolveTableSpec` as a separate rule.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,73 @@
+        (key, null)
+      case (key: String, value: Expression) =>
+        val newValue: String = try {
+          constantFold(value) match {
```
Shall we just use `value.eval()` since it is foldable?

dtenedor replied:
Sounds good, this is done.

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Hi @gengliangwang, thanks again for your reviews, this is ready again :)

dtenedor reviewed (commented):
Hi @gengliangwang, thanks again for your reviews, this is ready again :)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,89 @@
+            val dt = value.dataType
+            value match {
+              case Literal(null, _) =>
+                "null"
```
should it be `null` or `"null"`?

dtenedor replied:
It should be "null", this part computes the string value for the table option.

gengliangwang replied:
But string can be null too. If it is `"null"`, then the literal should be `Literal("null", StringType)`

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -437,6 +437,12 @@ trait V2CreateTablePlan extends LogicalPlan {
 }
 
+/** A trait used for logical plan nodes that create or replace new table specifications. */
+trait HasTableSpec extends LogicalPlan {
```
I wonder why we create a new trait here. Shall we simply make `UnresolvedOptionsList` part of`UnresolvedTableSpec`?

dtenedor replied:
I had originally created the trait to help with matching in the `ResolveTableSpec` rule. It's not strictly needed; I removed the trait now and updated the rule to inspect relevant fields instead.

I originally made `UnresolvedOptionsList` part of `UnresolvedTableSpec` :) but you suggested moving it out, which was helpful:

![image](https://github.com/apache/spark/assets/99207096/bce91bad-c5de-4f9a-9b76-165b7c037dba)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -437,6 +437,12 @@ trait V2CreateTablePlan extends LogicalPlan {
+/** A trait used for logical plan nodes that create or replace new table specifications. */
+trait HasTableSpec extends LogicalPlan {
+  def tableSpec: TableSpec
+  def unresolvedOptionsList: UnresolvedOptionsList
```
Even when a logical plan like `CreateTable` is resolved, it will carry the method `unresolvedOptionsList`

dtenedor replied:
Good point, I renamed this to `optionsListExpressions` to clarify.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Hi @gengliangwang I responded to your comments. I think this is ready to merge now, if you agree.

dtenedor committed:
respond to code review comments

dtenedor committed:
Merge branch 'master' into expression-properties
fetch from master

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor commented:
(Note, this is passing all CI again.)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableSpec.scala:
```diff
@@ -0,0 +1,90 @@
+            val dt = value.dataType
+            value match {
+              case Literal(null, _) =>
+                "null"
```
```suggestion
                null
```

dtenedor replied:
Sounds good, done.

gengliangwang reviewed (approved):
LGTM except for the last comment. Sorry for the slow response, I was focusing on another project.

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor commented:
> LGTM except for the last comment. Sorry for the slow response, I was focusing on another project.

No worries, thanks for the careful reviews :)

gengliangwang commented:
This has been iterated for many rounds. I will make a follow-up PR to further make the code simpler.
Merging to master.
